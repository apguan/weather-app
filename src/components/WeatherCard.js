import React, { Component } from 'react'
import Modal from 'react-responsive-modal'

import { Card, CardDetail, ModalDetails, CardContent } from './../StyledComponents.js'

export default class WeatherCard extends Component {
  state = {
    open: false
  }

  modalView = () => this.setState({ open: !this.state.open })

  displayDate = () => {
    let date = this.props.date.split('/').slice(0, 2).join('/')
    let localDate = new Date().toLocaleDateString().split('/').slice(0, 2).join('/')

    if (date === localDate) return 'Today'

    return date
  }

  getMostCommonWeatherCondition = (value) => {
    //value are the keys for the weather object.
    //can be used to get 'icons', 'main', or 'descriptions'
    let counter = this.props.conditions.reduce((acc, curr) => {
      if (curr[value] in acc) {
        acc[curr[value]]++;
      } else {
        acc[curr[value]] = 1;
      }
      return acc;
    }, {});

    let condition = Object.keys(counter).reduce((a, b) => counter[a] > counter[b] ? a : b);

    return condition;
  }

  render() {
    const { open } = this.state
    const { high, low, hourly } = this.props

    return (
      <div>
        {
          open ?
            <Modal open={open} onClose={this.modalView} center>
              {
                hourly.map((data, idx) => {
                  let time = data.dt_txt.split(',')[1]
                  return (
                    <ModalDetails key={idx}>
                      <p>time: {time}</p>
                      <span>temperature: {Math.ceil(data.main.temp)}</span>
                      <span>humidity: {data.main.humidity}</span>
                      <span>description: {data.weather[0].description}</span>
                    </ModalDetails>
                  )
                })
              }
            </Modal>
            :
            <Card onClick={this.modalView}>
              <h3>{this.displayDate()}</h3>

              <CardContent>
                <img src={`http://openweathermap.org/img/w/${this.getMostCommonWeatherCondition('icon')}.png`}></img>
                <p>{this.getMostCommonWeatherCondition('description')}</p>
              </CardContent>

              <CardDetail>
                <h4>high: {high}°</h4>
                <h4>low: {low}°</h4>
              </CardDetail>
            </Card>
        }
      </div>
    )
  }
}
