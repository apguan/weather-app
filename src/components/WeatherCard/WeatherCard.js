import React, { Component } from 'react'
import Modal from 'react-responsive-modal'
import ReactAnimatedWeather from 'react-animated-weather'

import DetailedWeather from '../DetailedWeather/DetailedWeather'

import { Card, CardDetail, CardContent, Title, WeatherLineItem } from '../../StyledComponents.js'

export default class WeatherCard extends Component {
  state = {
    open: false
  }

  modalView = () => this.setState({ open: !this.state.open })

  displayDate = () => {
    let date = this.props.date.split('/').slice(0, 2).join('/')
    let localDate = new Date().toLocaleDateString().split('/').slice(0, 2).join('/')

    if (date === localDate) return 'Today'

    return date
  }

  getMostCommonWeatherCondition = (value) => {
    //value are the keys for the weather object.
    //can be used to get 'icons', 'main', or 'descriptions'
    let counter = this.props.conditions.reduce((acc, curr) => {
      if (curr[value] in acc) {
        acc[curr[value]]++;
      } else {
        acc[curr[value]] = 1;
      }
      return acc;
    }, {});

    let condition = Object.keys(counter).reduce((a, b) => counter[a] > counter[b] ? a : b);

    return condition;
  }

  weatherIcon = () => {
    let map = {
      'Clear': 'CLEAR_DAY',
      'Clouds': 'PARTLY_CLOUDY_DAY',
      'Snow': 'SNOW',
      'Rain': 'RAIN',
      'Drizzle': 'RAIN',
      'Thunderstorm': 'SLEET',
      'Smoke': 'FOG',
      'Haze': 'FOG',
      'Dust': 'FOG'
    }
    let icon = map[this.getMostCommonWeatherCondition('main')] || 'CLEAR_DAY'

    return (
      <ReactAnimatedWeather
        icon={icon}
        color={'orange'}
        animate={true}
      />
    )
  }

  render() {
    const { high, low, hourly } = this.props

    return (
      <>
        <Modal
          open={this.state.open}
          onClose={this.modalView}
          closeIconSize={16}
          center
        >
          <Title>3-Hourly Forecast</Title>
          <DetailedWeather hourly={hourly} />
          <WeatherLineItem>
            <thead>
              <tr>
                <th>Time</th>
                <th>Temperature (F째)</th>
                <th>Humidity %</th>
                <th>Description</th>
              </tr>
            </thead>
            <tbody>
              {
                hourly.map((val, idx) => {
                  return (
                    <tr key={idx}>
                      <td>{val.dt_txt.split(',')[1]}</td>
                      <td>{Math.ceil(val.main.temp)}째</td>
                      <td>{val.main.humidity}%</td>
                      <td>{val.weather[0].description}</td>
                    </tr>
                  )
                })
              }
            </tbody>
          </WeatherLineItem>
        </Modal>

        <Card onClick={this.modalView}>
          <h3>{this.displayDate()}</h3>
          <CardContent>
            {this.weatherIcon()}
            <p>{this.getMostCommonWeatherCondition('description')}</p>
          </CardContent>

          <CardDetail>
            <h4>H: {high}째F</h4>
            <h4>L: {low}째F</h4>
          </CardDetail>
        </Card>
      </>
    )
  }
}
