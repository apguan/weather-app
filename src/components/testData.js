export const testData = [{ "dt": 1566723600, "main": { "temp": 58.66, "temp_min": 58.66, "temp_max": 58.66, "pressure": 1012.93, "sea_level": 1012.93, "grnd_level": 1005.85, "humidity": 82, "temp_kf": 0 }, "weather": [{ "id": 800, "main": "Clear", "description": "clear sky", "icon": "01n" }], "clouds": { "all": 0 }, "wind": { "speed": 4.34, "deg": 254.74 }, "sys": { "pod": "n" }, "dt_txt": "8/25/2019, 2:00:00 AM" }, { "dt": 1566734400, "main": { "temp": 58.19, "temp_min": 58.19, "temp_max": 58.19, "pressure": 1013.7, "sea_level": 1013.7, "grnd_level": 1006.39, "humidity": 83, "temp_kf": 0 }, "weather": [{ "id": 800, "main": "Clear", "description": "clear sky", "icon": "01n" }], "clouds": { "all": 0 }, "wind": { "speed": 2.86, "deg": 268.476 }, "sys": { "pod": "n" }, "dt_txt": "8/25/2019, 5:00:00 AM" }, { "dt": 1566745200, "main": { "temp": 60.71, "temp_min": 60.71, "temp_max": 60.71, "pressure": 1014.43, "sea_level": 1014.43, "grnd_level": 1007.05, "humidity": 77, "temp_kf": 0 }, "weather": [{ "id": 800, "main": "Clear", "description": "clear sky", "icon": "01d" }], "clouds": { "all": 0 }, "wind": { "speed": 1.83, "deg": 268.324 }, "sys": { "pod": "d" }, "dt_txt": "8/25/2019, 8:00:00 AM" }, { "dt": 1566756000, "main": { "temp": 69.02, "temp_min": 69.02, "temp_max": 69.02, "pressure": 1014.8, "sea_level": 1014.8, "grnd_level": 1007.49, "humidity": 62, "temp_kf": 0 }, "weather": [{ "id": 800, "main": "Clear", "description": "clear sky", "icon": "01d" }], "clouds": { "all": 0 }, "wind": { "speed": 6.49, "deg": 176.086 }, "sys": { "pod": "d" }, "dt_txt": "8/25/2019, 11:00:00 AM" }, { "dt": 1566766800, "main": { "temp": 74.54, "temp_min": 74.54, "temp_max": 74.54, "pressure": 1013.92, "sea_level": 1013.92, "grnd_level": 1006.29, "humidity": 46, "temp_kf": 0 }, "weather": [{ "id": 800, "main": "Clear", "description": "clear sky", "icon": "01d" }], "clouds": { "all": 0 }, "wind": { "speed": 11.01, "deg": 206.362 }, "sys": { "pod": "d" }, "dt_txt": "8/25/2019, 2:00:00 PM" }, { "dt": 1566777600, "main": { "temp": 74.49, "temp_min": 74.49, "temp_max": 74.49, "pressure": 1012.79, "sea_level": 1012.79, "grnd_level": 1004.97, "humidity": 41, "temp_kf": 0 }, "weather": [{ "id": 800, "main": "Clear", "description": "clear sky", "icon": "01n" }], "clouds": { "all": 0 }, "wind": { "speed": 10.85, "deg": 229.124 }, "sys": { "pod": "n" }, "dt_txt": "8/25/2019, 5:00:00 PM" }, { "dt": 1566788400, "main": { "temp": 65.57, "temp_min": 65.57, "temp_max": 65.57, "pressure": 1012.7, "sea_level": 1012.7, "grnd_level": 1004.85, "humidity": 59, "temp_kf": 0 }, "weather": [{ "id": 800, "main": "Clear", "description": "clear sky", "icon": "01n" }], "clouds": { "all": 0 }, "wind": { "speed": 8.55, "deg": 238.06 }, "sys": { "pod": "n" }, "dt_txt": "8/25/2019, 8:00:00 PM" }, { "dt": 1566799200, "main": { "temp": 61.12, "temp_min": 61.12, "temp_max": 61.12, "pressure": 1013.64, "sea_level": 1013.64, "grnd_level": 1006.1, "humidity": 67, "temp_kf": 0 }, "weather": [{ "id": 800, "main": "Clear", "description": "clear sky", "icon": "01n" }], "clouds": { "all": 0 }, "wind": { "speed": 6.4, "deg": 228.335 }, "sys": { "pod": "n" }, "dt_txt": "8/25/2019, 11:00:00 PM" }]
export const testDate = "8/21/2019"
export const testConditions = [{ "id": 800, "main": "Clear", "description": "clear sky", "icon": "01n" }, { "id": 800, "main": "Clear", "description": "clear sky", "icon": "01n" }, { "id": 800, "main": "Clear", "description": "clear sky", "icon": "01d" }, { "id": 800, "main": "Clear", "description": "clear sky", "icon": "01d" }, { "id": 800, "main": "Clear", "description": "clear sky", "icon": "01d" }, { "id": 800, "main": "Clear", "description": "clear sky", "icon": "01n" }, { "id": 800, "main": "Clear", "description": "clear sky", "icon": "01n" }, { "id": 800, "main": "Clear", "description": "clear sky", "icon": "01n" }]