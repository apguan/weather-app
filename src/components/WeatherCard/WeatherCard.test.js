import React from 'react'
import { shallow, mount, render } from 'enzyme';

import WeatherCard from './WeatherCard'
import { testData, testDate, testConditions } from '../testData'

it('renders incorrectly when there are no items', () => {
  const wrapper = shallow(<WeatherCard date={testDate} hourly={testData} conditions={testConditions} />)
  expect(wrapper).toMatchSnapshot()
})