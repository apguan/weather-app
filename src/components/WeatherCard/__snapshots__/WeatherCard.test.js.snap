// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders incorrectly when there are no items 1`] = `ShallowWrapper {}`;
